### PR TITLE
`Development`: Replace grading criterion list with set

### PR DESCRIPTION
**Pull request by b-fein**
@coderabbitai ignore

#### General

- [ ] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I documented the Java code using JavaDoc style.

### Motivation and Context
It does not make sense to have duplicated `GradingCriteria` in an exercise and duplicated `GradingInstructions` in a criterion as already noted in a `todo` comment.

### Description
This PR replaces the lists with sets to ensure uniqueness.

This also unblocks other PRs:
- #7928
- #7725

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 exercise for each type (feel free to only test some of them; if so, please comment which ones you tested it with) that has at least one submission and grading criteria enabled:
  - 1 Programming Exercise with manual assessment
  - 1 Modelling exercise
  - 1 Text exercise
  - 1 File upload exercise

Anything related to the manual assessment using grading criteria could be affected. Some pointers for items that should be tested:

- Create/edit/delete grading criteria and instructions in the exercise settings.
- Assess a submission using the grading criteria (drag-and-drop from the right sidebar into feedback).
- Update an assessment that uses grading instructions on feedback.
- Delete feedback that has a grading instruction.
- Edit the grading instructions after an assessment was already created in the exercise.

#### Exam Mode Testing
As far as I am aware, there is no special handling of grading criteria in case of exam exercises. Therefore, I don’t think special testing is required. The testing steps above apply same as to a course exercise.

#### Performance Review
- [ ] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage
All production code that was covered before is covered now as well since there are only trivial replacements from `List` to `Set` without any required logic changes.
Code review should ensure, though, that the updated assertions still make sense.

### CODE REVIEW

File: src/test/java/de/tum/in/www1/artemis/assessment/ResultServiceIntegrationTest.java [hunk 1]
```diff
@@ -367,16 +368,18 @@ else if (i % 4 == 0) {
...
+        final GradingCriterion criterion2 = GradingCriterionUtil.findGradingCriterionByTitle(fileUploadExercise, "test title2");
 
-        final GradingInstruction instruction1a = criterion1.getStructuredGradingInstructions().get(0);
-        final GradingInstruction instruction1b = criterion1.getStructuredGradingInstructions().get(1);
-        final GradingInstruction instruction2 = criterion2.getStructuredGradingInstructions().get(0);
+        final List<GradingInstruction> gradingInstructions1 = List.copyOf(criterion1.getStructuredGradingInstructions());
+        final GradingInstruction instruction1a = gradingInstructions1.get(0);
+        final GradingInstruction instruction1b = gradingInstructions1.get(1);
+        final List<GradingInstruction> gradingInstructions2 = List.copyOf(criterion2.getStructuredGradingInstructions());
+        final GradingInstruction instruction2 = gradingInstructions2.get(0);
```
Comment (Strohgelaender) [inline src/test/java/de/tum/in/www1/artemis/assessment/ResultServiceIntegrationTest.java:382]: Is the ordering fixed when using `copyOf` on a Set?

File: src/test/java/de/tum/in/www1/artemis/exercise/ExerciseUtilService.java [hunk 2]
```diff
@@ -113,24 +112,25 @@ public Exercise addMaxScoreAndBonusPointsToExercise(Exercise exercise) {
      * @param exercise The exercise to which the grading instructions should be added.
      * @return The list of grading criteria.
      */
-    public List<GradingCriterion> addGradingInstructionsToExercise(Exercise exercise) {
+    public Set<GradingCriterion> addGradingInstructionsToExercise(Exercise exercise) {
         GradingCriterion emptyCriterion = ExerciseFactory.generateGradingCriterion(null);
         List<GradingInstruction> instructionWithNoCriteria = ExerciseFactory.generateGradingInstructions(emptyCriterion, 1, 0);
```
Comment (Strohgelaender) [inline src/test/java/de/tum/in/www1/artemis/exercise/ExerciseUtilService.java:117]: This method should also return a Set.

File: src/test/java/de/tum/in/www1/artemis/exercise/modelingexercise/ModelingExerciseIntegrationTest.java [hunk 3]
```diff
@@ -123,14 +124,7 @@ void testGetModelingExercise_asTA() throws Exception {
...
 
         ModelingExercise receivedModelingExercise = request.get("/api/modeling-exercises/" + classExercise.getId(), HttpStatus.OK, ModelingExercise.class);
-        gradingCriteria = exerciseUtilService.addGradingInstructionsToExercise(receivedModelingExercise);
-        assertThat(receivedModelingExercise.getGradingCriteria().get(0).getTitle()).isNull();
-        assertThat(receivedModelingExercise.getGradingCriteria().get(1).getTitle()).isEqualTo("test title");
-
-        assertThat(gradingCriteria.get(0).getStructuredGradingInstructions()).hasSize(1);
-        assertThat(gradingCriteria.get(1).getStructuredGradingInstructions()).hasSize(3);
-        assertThat(gradingCriteria.get(0).getStructuredGradingInstructions().get(0).getInstructionDescription())
-                .isEqualTo("created first instruction with empty criteria for testing");
```
Comment (Strohgelaender) [inline src/test/java/de/tum/in/www1/artemis/exercise/modelingexercise/ModelingExerciseIntegrationTest.java:133]: Why did you remove these assertions?

Comment (b-fein) [reply src/test/java/de/tum/in/www1/artemis/exercise/modelingexercise/ModelingExerciseIntegrationTest.java:133]: Those only test that our utility (`exerciseUtilService.addGradingInstructionsToExercise`) can create some grading criteria. There is no more actual interaction with the production code.

The fact that the utility produces criteria and instructions is extensively tested by using the results in other tests.

File: src/test/java/de/tum/in/www1/artemis/exercise/ExerciseUtilService.java (hunk 2 above)
Comment (b-fein) [reply src/test/java/de/tum/in/www1/artemis/exercise/ExerciseUtilService.java:117]: IIRC there were a few cases where the caller then further adapted each item in the list with some additional changes. This is of course easier when it remains a list where you can directly index items. I’ll check again, though.

File: src/test/java/de/tum/in/www1/artemis/assessment/ResultServiceIntegrationTest.java (hunk 1 above)
Comment (b-fein) [reply src/test/java/de/tum/in/www1/artemis/assessment/ResultServiceIntegrationTest.java]: No, but it only contains one element anyway. Made that clear in 580ea3a5ea6652446ca0a29ffcb8e5b876c1b847 with an additional assertion on the size.

File: src/test/java/de/tum/in/www1/artemis/exercise/ExerciseUtilService.java (hunk 2 above)
Comment (b-fein) [reply src/test/java/de/tum/in/www1/artemis/exercise/ExerciseUtilService.java]: Apparently not any more after the other refactorings. Changed to a set in e430ce60b99b847a13bb5dcba691b5e2bf960078.

File: src/test/java/de/tum/in/www1/artemis/assessment/ResultServiceIntegrationTest.java (hunk 1 above)
Comment (Strohgelaender) [reply src/test/java/de/tum/in/www1/artemis/assessment/ResultServiceIntegrationTest.java]: The issue remains for `instruction1a` and `instruction1b` above.

Comment (b-fein) [reply src/test/java/de/tum/in/www1/artemis/assessment/ResultServiceIntegrationTest.java]: → 48bbbd0ea966d03eea4fc3cbb1ae050f77480f81

**Review by Gusti2010 — APPROVED** [review]
Manually tested on ts11_Legacy. Works as expected for text, file upload and modeling exercises!

**Review by AntonGeTUM — APPROVED** [review]
Manually tested with programming, text and modelling exercises, works as expected

**Review by georgyia — APPROVED** [review]
manually tested on ts2, works as expected

**Review by chrisknedl — APPROVED** [review]
Code looks good to me. Thanks for taking the time to fix this, there was apparently a lot to adapt regarding the tests.

**Review by coolchock — APPROVED** [review]
code lgtm
